bootstrap/cmd/bootstrap/app/gcpUtils.go

```diff
@@ -83,4 +104,98 @@ func (s *ksServer)GetDeploymentStatus(ctx context.Context, req CreateRequest) (s
 		return "", err
 	}
 	return dm.Operation.Status, nil
-}
\ No newline at end of file
+}
+
+func GetUpdatedPolicy(currentPolicy *cloudresourcemanager.Policy, iamConf *IamConf, req ApplyIamRequest) cloudresourcemanager.Policy {
+	// map from role to members.
+	policyMap := map[string]map[string]bool {}
+	for _, binding := range currentPolicy.Bindings {
+		policyMap[binding.Role] = make(map[string]bool)
+		for _, member := range binding.Members {
+			policyMap[binding.Role][member] = true
+		}
+	}
+
+	// Replace placeholder with actual identity.
+	saMapping := map[string]string {
+		"set-kubeflow-admin-service-account": fmt.Sprintf("serviceAccount:%v-admin@%v.iam.gserviceaccount.com", req.Cluster, req.Project),
+		"set-kubeflow-user-service-account": fmt.Sprintf("serviceAccount:%v-user@%v.iam.gserviceaccount.com", req.Cluster, req.Project),
+		"set-kubeflow-vm-service-account": fmt.Sprintf("serviceAccount:%v-vm@%v.iam.gserviceaccount.com", req.Cluster, req.Project),
+		"set-kubeflow-iap-account": fmt.Sprintf("user:%v", req.Email),
+	}
+	for _, binding := range iamConf.IamBindings {
+		for _, member := range binding.Members {
+			actualMember := member
+			if val, ok := saMapping[member]; ok {
+				actualMember = val
+			}
+			for _, role := range binding.Roles {
+				if req.Action == "add" {
+					policyMap[role][actualMember] = true
+				} else {
+					// action == "remove"
+					policyMap[role][actualMember] = false
+				}
+			}
+		}
+	}
+	newPolicy := cloudresourcemanager.Policy{}
+	for role, memberSet := range policyMap {
+		binding := cloudresourcemanager.Binding{}
+		binding.Role = role
+		for member, exists := range memberSet {
+			if exists {
+				binding.Members = append(binding.Members, member)
+			}
+		}
+		newPolicy.Bindings = append(newPolicy.Bindings, &binding)
+	}
+	return newPolicy
+}
+
+func (s *ksServer)ApplyIamPolicy(ctx context.Context, req ApplyIamRequest) error {
```

Comment:
Let's add retry into ApplyIamPolicy.
There's a small chance that concurrent iam change might occur for same project.

Reply:
Done, PTAL